ush/set_cycle_and_obs_timeinfo.py

```diff
@@ -0,0 +1,736 @@
+        else:
+            consec_obs_days_lists.append([day_crnt])
+        # Update the value of the previous day in preparation for the next
+        # iteration of the loop.
+        day_prev = day_crnt
+
+    # Use the list of lists of consecutive obs days to construct a list of
+    # ROCOTO-style cycledef strings that each represent a set of consecutive
+    # obs days when included in a <cycledef> tag in a ROCOTO XML.  Each
+    # string in this new list corresponds to a series of consecutive days on
+    # which observations are needed (where by "consecutive" we mean no days
+    # are skipped), and there is at least a one-day gap between each such
+    # series.  These cycledefs together represent all the days (i.e. over all
+    # cycles of the experiment) on which observations are needed.
+    cycledefs_all_obs_days = list()
+    for consec_obs_days_list in consec_obs_days_lists:
+        cycledef_start = consec_obs_days_list[0].strftime('%Y%m%d%H%M')
+        cycledef_end = consec_obs_days_list[-1].strftime('%Y%m%d%H%M')
+        cycledefs_all_obs_days.append(' '.join([cycledef_start, cycledef_end, '24:00:00']))
+
+    return cycledefs_all_obs_days
+
+
+def get_obs_retrieve_times_by_day(
+    vx_config, fcst_output_times_all_cycles, obs_days_all_cycles):
+    """
+    This function generates dictionary of dictionaries that, for each
+    combination of obs type needed and each obs day, contains a string list
+    of the times at which that type of observation is needed on that day.
+    The elements of each list are formatted as 'YYYYMMDDHH'.
+
+    Args:
+        vx_config:
+        The verification configuration dictionary.
+
+        fcst_output_times_all_cycles:
+        Dictionary containing a list of forecast output times over all cycles for
+        instantaneous fields and a second analogous list for cumulative fields.
+        Each element of these lists is a string of the form 'YYYYMMDDHH'.
+
+        obs_days_all_cycles:
+        Dictionary containing a list of observation days (i.e. days on which
+        observations are needed to perform verification) over all cycles for
+        instantaneous fields and a second analogous list for cumulative fields.
+        Each element of these lists is a string of the form 'YYYYMMDD'.
+
+    Returns:
+        obs_retrieve_times_by_day:
+        Dictionary of dictionaries containing times at which each type of obs is
+        needed on each obs day.
+    """
+    # Convert string contents of input dictionaries to datetime objects.
+    for time_type in ['cumul', 'inst']:
+      fcst_output_times_all_cycles[time_type] \
+      = [datetime.strptime(fcst_output_times_all_cycles[time_type][i], "%Y%m%d%H")
+                           for i in range(len(fcst_output_times_all_cycles[time_type]))]
+      obs_days_all_cycles[time_type] \
+      = [datetime.strptime(obs_days_all_cycles[time_type][i], "%Y%m%d")
+                           for i in range(len(obs_days_all_cycles[time_type]))]
```

Comment:
Indentation should be 4 spaces per level
```suggestion
    for time_type in ['cumul', 'inst']:
        fcst_output_times_all_cycles[time_type] \
        = [datetime.strptime(fcst_output_times_all_cycles[time_type][i], "%Y%m%d%H")
                           for i in range(len(fcst_output_times_all_cycles[time_type]))]
        obs_days_all_cycles[time_type] \
      = [datetime.strptime(obs_days_all_cycles[time_type][i], "%Y%m%d")
                           for i in range(len(obs_days_all_cycles[time_type]))]
```